### ISSUE TITLE
systemd is heavily IO-limited reading unit files at boot

### ISSUE THREAD

**Issue by Jookia** [systemd/systemd#16736]
**systemd version the issue has been seen with**
245

**Used distribution**
 Buildroot 2020.05.1

**Expected behaviour you didn't see**
 systemd finds unit files for sysinit.target and boots them as soon as possible

**Unexpected behaviour you saw**
 systemd finds ALL unit files on the drive and parses them before booting, wasting a second of boot time to build its execution graph.

From grabserial with a debug system:

```
[3.981441 0.004190] [    2.753955] systemd[1]: Successfully forked off '(sd-executor)' as PID 89.
... systemd running generators
... pages of 'systemd[1]: unit_file_build_name_map: '
[6.343576 0.492643] [  OK  ] Created slice system-getty.slice.
```

bootchart confirms the issue, though it's faster without debug: systemd takes around a second to load unit files before starting the system.

I've attached plots and logs as a zip file: [bootcharts.zip](https://github.com/systemd/systemd/files/5078399/bootcharts.zip)

Ideally systemd could only read units for a certain target before booting.

**Steps to reproduce the problem**
Set up bootchart on a low-end Linux system (ARM with slow eMMC) and try to boot systemd.

**Comment by poettering** [comment]
This is not what happens. systemd reads all unit files referenced by units it is requested to boot into, i.e. all transitive depencies of default.target, and that's all.

**Comment by Jookia** [comment]
Are you sure? Because booting with unit=rescue.target still reads unit
files for things like bluetooth, wpa_supplicant, etc.

Is there some example that boots a bash shell and no other services at
all?

**Comment by Jookia** [comment]
Here's the debug log when booting with unit=rescue.target. It takes about the same time (0.05s longer) to load all the units.

[linuxboot_rescue.txt](https://github.com/systemd/systemd/files/5081228/linuxboot_rescue.txt)

**Comment by poettering** [comment]
> Are you sure? Because booting with unit=rescue.target still reads unit files for things like bluetooth, wpa_supplicant, etc.

Probably because the units are referenced in some way, maybe through Conflicts= or so.

**Comment by poettering** [comment]
> Here's the debug log when booting with unit=rescue.target. It takes about the same time (0.05s longer) to load all the units.
> 
> [linuxboot_rescue.txt](https://github.com/systemd/systemd/files/5081228/linuxboot_rescue.txt)

What is that supposed to show?

Note that unit_file_build_name_map() iterates through the dirs, but opens no files. It just builds a map where to find what, tracks down some symlinks, to accelrate later lookups, so that we don#t need to search all the time when we need to actually open a file.

I don#t think there's anything here to fix.

How did you determine that PID 1 "reads" files? just by the log output? or did you strace things or so? If the former I figure there's some confusion on when systemd actually opens files, or when it just stats them

**Comment by Jookia** [comment]
I guess I assumed it was reading them because it's taking a long time to
iterate through the directories and build this map. I guess it's not
reading them, I haven't straced things.

The boot log is showing that it's taking a second to do all this.

**Comment by vcaputo** [comment]
Well on a slow storage device with nothing cached in a cold boot even just traversing a tree for metadata can take some time.

You might be able to get better cold metadata performance from a different filesystem and/or by tuning mkfs options.

**Comment by Jookia** [comment]
I spent a while debugging this and found that a lot of time systemd is parsing unit files from disk. It would be good if the system could load serialized units at boot instead of parsing them, but systemd doesn't have that feature. I don't know how long this takes anyway, I don't know how much that would speed it up, it's painful to get concrete numbers on an embedded system. The "I don't think there's anything here to fix." response seems to imply having the system hang for a second during boot reading files isn't a huge problem, and if that's the case I'll close the issue and not bother any more about this.

**Comment by poettering** [comment]
I mean, optimizations are always welcome. But reading files off disk that we need isn't really a "bug that needs fixing".

I am not a big fan of cache schemes I must say (i.e. where we compile a stateful cache from the individual user and vendor provided files in order to maybe improve access time because having many files is often times slower than having a single one), because they get easily out of sync, and we generate units/symlinks dynamically at boot through the generator and preset logic.

before doing anything like that, maybe some lower hanging fruit would be to somehow tell the kernel to read in all files ahead of time (i.e. fork off a thread or subprocess that loads a directory into memory by going through it and doing fadvise() on each file, in the background while the main program stays as it is and reads things one by one parsing them, in which case they are hopefully all already cached in). Or maybe use io_uring to schedule all reads at once and let the kernel figure out the order in which it wants to read that. i.e. doing an on-disk cache that can get out of sync easily is something I am not fond of, but doing optimizations with the unit files as they are via better Linux API usage is very sympathetic to me.

Either way, we'd probably need a contributed PR with these optimizations, it's unlikely that from upstream we'll work on this, as our systems are not that IO constrained typiclly.

**Comment by Jookia** [comment]
Yeah after a spending a while thinking about this, it might be that the best solution here would be to try attack the issue at that layer. If I get anything interesting done on this level I'll update this thread.
